jni/src/Foreign/JNI.hs

```diff
@@ -197,11 +203,28 @@ instance Exception ArrayCopyFailed
 -- | Map Java exceptions to Haskell exceptions.
 throwIfException :: Ptr JNIEnv -> IO a -> IO a
 throwIfException env m = m `finally` do
-    J excptr <- [CU.exp| jthrowable { (*$(JNIEnv *env))->ExceptionOccurred($(JNIEnv *env)) } |]
+    excptr <- [CU.exp| jthrowable { (*$(JNIEnv *env))->ExceptionOccurred($(JNIEnv *env)) } |]
     unless (excptr == nullPtr) $ do
       [CU.exp| void { (*$(JNIEnv *env))->ExceptionDescribe($(JNIEnv *env)) } |]
       [CU.exp| void { (*$(JNIEnv *env))->ExceptionClear($(JNIEnv *env)) } |]
-      throwIO $ JavaException (J excptr)
+      newGlobalJ env excptr >>= throwIO . JavaException
+
+-- | Creates a global reference to a given Java value.
+--
+-- Arranges for 'deleteGlobalRef' to be called when the global reference
+-- is no longer reachable on the Haskell side.
+newGlobalJ :: Ptr JNIEnv -> Ptr (J ty) -> IO (J ty)
+newGlobalJ env (castPtr -> obj) = do
+    g <- [CU.exp| jobject {
+      (*$(JNIEnv *env))->NewGlobalRef($(JNIEnv *env),
+                                      $(jobject obj)) } |]
+    fmap (unsafeCast . J) $ fixIO $ newConcForeignPtr g . finalize
+  where
+    finalize fp = withJNIEnv $ \env ->
+      withForeignPtr fp $ \objp ->
+      [CU.block| void {
+        (*$(JNIEnv *env))->DeleteGlobalRef($(JNIEnv *env),
+                                           $(jobject objp)); } |]
```

Comment:
I have observed this in practice. At any rate, I believe that's what the following error is essentially saying:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f99995af137, pid=9171, tid=0x00007f999a15e9c0
#
# JRE version: OpenJDK Runtime Environment (8.0_122) (build 1.8.0_122-04)
# Java VM: OpenJDK 64-Bit Server VM (25.122-b04 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# V  [libjvm.so+0x5ed137]  ThreadStateTransition::transition_from_native(JavaThread*, JavaThreadState)+0x17
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/mboes/src/tweag/inline-java/jvm/hs_err_pid9171.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#                                                                                       
Completed 2 action(s).
Test suite failure for package jvm-0.1.2
    spec:  exited with: ExitFailure (-6)
Logs printed to console
```

It only happens once all tests pass and the program is about to terminate.